src/setup_node_env/node_version_validator.js

```diff
@@ -28,27 +28,201 @@
+ */
+var requiredParts = pkgEngineNodeVersion.match(
+  /^\s*(>=?|\^|~|=)?\s*(\d+)(?:\.(\d+))?(?:\.(\d+))?(?:[^\d.].*)?$/
+);
+if (requiredParts === null) {
+  console.error(
+    'OpenSearch Dashboards did not report its required version of the Node.js runtime in a valid ' +
+      'format. Please revert any changes that might have been made to the package.json file and try again.'
+  );
+  process.exit(23);
+}
+
+var comparatorVersion = {
+  major: requiredParts[2],
+  minor: requiredParts[3],
+  patch: requiredParts[4],
+};
+var comparatorOperator = requiredParts[1] || '=';
+var rangeBottom = {
+  major: parseInt(comparatorVersion.major, 10) || 0,
+  minor: parseInt(comparatorVersion.minor, 10) || 0,
+  patch: parseInt(comparatorVersion.patch, 10) || 0,
+  inclusive: comparatorOperator.indexOf('>') === -1 || comparatorOperator.indexOf('=') > -1,
+};
+var rangeTop = undefined;
+
+if (!rangeBottom.major) {
+  console.error(
+    'OpenSearch Dashboards did not report its required version of the Node.js runtime in a valid ' +
+      'format. Please revert any changes that might have been made to the package.json file and try again.'
+  );
+  process.exit(24);
+}
+
+if (comparatorOperator === '>') {
+  if (!comparatorVersion.minor) {
+    // >3 is >=4.0.0
+    rangeBottom.major += 1;
+    rangeBottom.inclusive = true;
+  } else if (!comparatorVersion.patch) {
+    // >3.1 is >=3.2.0
+    rangeBottom.minor += 1;
+    rangeBottom.inclusive = true;
+  }
+}
+
+// =3 is ~3.0.0 and =3.1 is ~3.1.0
+if (comparatorOperator === '=' && (!comparatorVersion.minor || !comparatorVersion.patch)) {
+  comparatorOperator = '~';
+}
+
+// =3.1.4
+if (comparatorOperator === '=') {
+  rangeTop = {
+    major: rangeBottom.major,
+    minor: rangeBottom.minor,
+    patch: rangeBottom.patch,
+    inclusive: true,
+  };
+} else if (comparatorOperator === '~') {
+  if (comparatorVersion.minor) {
+    // ~3.1.4 and ~3.1 are <3.2.0
+    rangeTop = {
+      major: rangeBottom.major,
+      minor: rangeBottom.minor + 1,
+      patch: 0,
+    };
+  } else {
+    // ~3 is <4.0.0
+    rangeTop = {
+      major: rangeBottom.major + 1,
+      minor: 0,
+      patch: 0,
+    };
+  }
+} else if (comparatorOperator === '^') {
+  // ^3.1.4 is <4.0.0
+  rangeTop = {
+    major: rangeBottom.major + 1,
+    minor: 0,
+    patch: 0,
+  };
+}
+
+function getVersionCompatibilityMessage() {
+  var versionBottom = 'v' + rangeBottom.major + '.' + rangeBottom.minor + '.' + rangeBottom.patch;
+  if (comparatorOperator === '=') {
+    return 'Please use Node.js ' + versionBottom + '.';
+  }
+
+  var message =
+    'Please use a Node.js runtime version that is ' +
+    (rangeBottom.inclusive ? versionBottom + ' or greater' : 'greater than ' + versionBottom);
+
+  if (!rangeTop) return message + '.';
+
+  var versionTop = 'v' + rangeTop.major + '.' + rangeTop.minor + '.' + rangeTop.patch;
+  // The only operator with a truey `rangeTop.inclusive` is the `=` which was handled above
```

Comment:
nit - "truthy" seems the more common term
```suggestion
  // The only operator with a truthy `rangeTop.inclusive` is the `=` which was handled above
```

Reply:
You know how long I sat thinking about this? I knew it didn't sound right to me.

Reply:
Done.